src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -176,6 +179,20 @@ def _make_candidate_from_link(
         name: Optional[NormalizedName],
         version: Optional[CandidateVersion],
     ) -> Optional[Candidate]:
+        base: Optional[BaseCandidate] = self._make_base_candidate_from_link(
+            link, template, name, version
+        )
+        if not extras or base is None:
+            return base
+        return self._make_extras_candidate(base, extras, comes_from=template)
+
+    def _make_base_candidate_from_link(
```

Comment:
`_make_candidate_from_link` actually always returned a `BaseCandidate` when no extras where specified. The new extras handling requires a base candidate so I extracted this part.